File: app/models/screen.rb
# frozen_string_literal: true

# == Schema Information
#
# Table name: screens
#
#  id                     :integer          not null, primary key
#  description            :string
#  name                   :string
#  result                 :string
#  collaborator           :string
#  conditions             :string
#  requirements           :string
#  created_at             :datetime         not null
#  updated_at             :datetime         not null
#  deleted_at             :datetime
#  component_graph_data   :jsonb
#  plain_text_description :text
#
# Indexes
#
#  index_screens_on_deleted_at  (deleted_at)
#

class Screen < ApplicationRecord
  has_logidze
  acts_as_paranoid
  include ElementUIStateScopes
  include PgSearch::Model
  include Collectable
  include ElementCodes
  include Taggable
  include Labimotion::Segmentable

  serialize :description, Hash

  multisearchable against: [:name, :conditions, :requirements]

  # search related
  pg_search_scope :search_by_screen_name, against: :name
  pg_search_scope :search_by_conditions, against: :conditions
  pg_search_scope :search_by_requirements, against: :requirements
  pg_search_scope :search_by_substring, against: [:name, :conditions, :requirements],
                                        using: {trigram: {threshold:  0.0001}}

  scope :by_name, ->(query) { where('name ILIKE ?', "%#{sanitize_sql_like(query)}%") }
  scope :by_conditions, ->(query) { where('conditions ILIKE ?', "%#{sanitize_sql_like(query)}%") }
  scope :by_requirements, ->(query) { where('requirements ILIKE ?', "%#{sanitize_sql_like(query)}%") }
  scope :by_wellplate_ids, ->(ids) { joins(:wellplates).where('wellplates.id in (?)', ids) }
  scope :includes_for_list_display, -> { includes(:comments) }

  has_many :collections_screens, dependent: :destroy
  has_many :collections, through: :collections_screens

  has_many :attachments, as: :attachable, dependent: :nullify

  has_many :screens_wellplates, dependent: :destroy
  has_many :wellplates, through: :screens_wellplates

  has_many :research_plans_screens, dependent: :destroy
  has_many :research_plans, through: :research_plans_screens

  has_many :comments, as: :commentable, dependent: :destroy

  has_one :container, :as => :containable

  before_save :description_to_plain_text

  accepts_nested_attributes_for :collections_screens

  def analyses
    self.container ? self.container.analyses : []
  end

  def update_body_attachments(original_identifier, copy_identifier)
    remap_richtext_attachment_identifiers(original_identifier, copy_identifier)
    save!
  end

  def remap_richtext_attachment_identifiers(original_identifier, copy_identifier)
    remap_delta_op_identifiers(description, original_identifier, copy_identifier)
  end

  INLINE_BLOT_KEYS = %w[attachment-image attachment-file].freeze

  private

  def remap_delta_op_identifiers(delta, original_identifier, copy_identifier)
    ops = delta.is_a?(Hash) ? delta['ops'] : nil
    return unless ops.is_a?(Array)

    ops.each do |op|
      insert = op.is_a?(Hash) ? op['insert'] : nil
      remap_insert_blots(insert, original_identifier, copy_identifier) if insert.is_a?(Hash)
    end
  end

  def remap_insert_blots(insert, original_identifier, copy_identifier)
    INLINE_BLOT_KEYS.each do |blot_key|
      payload = insert[blot_key]
      next unless payload.is_a?(Hash) && payload['attachment_identifier'] == original_identifier

      payload['attachment_identifier'] = copy_identifier
    end
  end

  def description_to_plain_text
    return unless description_changed?

    self.plain_text_description = Chemotion::QuillToPlainText.convert(description)
  end
end

File: app/packs/src/components/models/Screen.js
import Element from './Element';
import Wellplate from './Wellplate';
import Container from './Container';

export default class Screen extends Element {
  static buildEmpty(collectionID) {
    const descriptionDefault = {
      ops: [{ insert: '' }]
    };

    return new Screen({
      collection_id: collectionID,
      type: 'screen',
      name: 'New Screen',
      collaborator: '',
      requirements: '',
      conditions: '',
      result: '',
      description: descriptionDefault,
      wellplates: [],
      research_plans: [],
      container: Container.init(),
    });
  }

  static buildFromWellplatesAndCollectionId(clipboardWellplates, collectionID) {
    const descriptionDefault = {
      ops: [{ insert: '' }]
    };

    return new Screen({
      collection_id: collectionID,
      type: 'screen',
      name: 'New Screen with Wellplates',
      collaborator: '',
      requirements: '',
      conditions: '',
      result: '',
      description: descriptionDefault,
      wellplates: clipboardWellplates,
      research_plans: [],
      container: Container.init(),
    });
  }

  serialize() {
    return super.serialize({
      name: this.name,
      collaborator: this.collaborator,
      result: this.result,
      conditions: this.conditions,
      requirements: this.requirements,
      description: this.description,
      wellplate_ids: this.wellplateIDs,
      research_plan_ids: this.researchPlanIDs,
      container: this.container,
    });
  }

  get name() {
    return this._name;
  }

  set name(name) {
    this._name = name;
  }

  get collaborator() {
    return this._collaborator;
  }

  set collaborator(collaborator) {
    this._collaborator = collaborator;
  }

  get requirements() {
    return this._requirements;
  }

  set requirements(requirements) {
    this._requirements = requirements;
  }

  get conditions() {
    return this._conditions;
  }

  set conditions(conditions) {
    this._conditions = conditions;
  }

  get result() {
    return this._result;
  }

  set result(result) {
    this._result = result;
  }

  get description() {
    return this._description;
  }

  set description(description) {
    this._description = description;
  }

  get wellplates() {
    return this._wellplates;
  }

  set wellplates(wellplates) {
    this._wellplates = wellplates.map(w => new Wellplate(w));
  }

  get wellplateIDs() {
    return this._wellplates.map(w => w.id);
  }

  get researchPlanIDs() {
    return this.research_plans.map(rp => rp.id);
  }

  title() {
    return this.name;
  }
}
